File: it-agents/skills/code-architect/references/code-examples/cpp/Iterator.cpp
// Iterator — traverse a collection without exposing its representation.
#include <iostream>
#include <memory>
#include <string>
#include <vector>

template <typename T>
class Iterator {
public:
    virtual ~Iterator() = default;
    virtual bool hasNext() const = 0;
    virtual const T& next() = 0;
};

template <typename T>
class Collection {
public:
    virtual ~Collection() = default;
    virtual std::unique_ptr<Iterator<T>> createIterator() const = 0;
};

// Ring buffer with a custom traversal order, hidden from the client.
class Ring : public Collection<std::string> {
    std::vector<std::string> items;
    size_t start;
public:
    explicit Ring(size_t startIndex) : start(startIndex) {}
    void add(const std::string& v) { items.push_back(v); }
    const std::vector<std::string>& raw() const { return items; }
    size_t startIndex() const { return start; }

    std::unique_ptr<Iterator<std::string>> createIterator() const override;
};

class RingIterator : public Iterator<std::string> {
    const Ring& ring;
    size_t step = 0;
public:
    explicit RingIterator(const Ring& r) : ring(r) {}
    bool hasNext() const override { return step < ring.raw().size(); }
    const std::string& next() override {
        size_t idx = (ring.startIndex() + step) % ring.raw().size();
        ++step;
        return ring.raw()[idx];
    }
};

std::unique_ptr<Iterator<std::string>> Ring::createIterator() const {
    return std::make_unique<RingIterator>(*this);
}

int main() {
    Ring ring{2};
    ring.add("a"); ring.add("b"); ring.add("c"); ring.add("d");
    auto it = ring.createIterator();
    while (it->hasNext()) std::cout << it->next() << " ";
    std::cout << "\n";
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/rust/iterator.rs
// Iterator — traverse a collection sequentially without exposing its internal
// representation, by implementing a custom Iterator.

// A simple ring buffer with a hidden layout.
struct RingBuffer {
    slots: Vec<i32>,
    start: usize,
}

impl RingBuffer {
    fn new(slots: Vec<i32>, start: usize) -> Self {
        RingBuffer { slots, start }
    }

    // Hand out an iterator instead of revealing the storage.
    fn iter(&self) -> RingIter<'_> {
        RingIter { buffer: self, pos: 0 }
    }
}

struct RingIter<'a> {
    buffer: &'a RingBuffer,
    pos: usize,
}

impl<'a> Iterator for RingIter<'a> {
    type Item = i32;

    fn next(&mut self) -> Option<i32> {
        if self.pos >= self.buffer.slots.len() {
            return None;
        }
        let idx = (self.buffer.start + self.pos) % self.buffer.slots.len();
        self.pos += 1;
        Some(self.buffer.slots[idx])
    }
}

fn main() {
    let ring = RingBuffer::new(vec![10, 20, 30, 40], 2);

    print!("traversal:");
    for value in ring.iter() {
        print!(" {}", value);
    }
    println!();

    // Works with standard adaptors because it's a real Iterator.
    let total: i32 = ring.iter().sum();
    println!("sum = {}", total);
}
